### PR TITLE
utils: parse quotes when splitting strings

### PR DESCRIPTION
**Pull request by tsaarni**
This change makes it possible to add configuration parameters with whitespace in them by adding support for quotes. For example, user might want to use `add` operation in modify filter to add value with whitespace in it:

```yaml
[FILTER]
    name modify
    match *
    condition key_value_matches message .*FOO.*
    add facility "audit log"
```

When processing the string, the parser will peek into the first non-separator character. If it is quote, it will expect quoted string and process until end-quote. Otherwise it will process the token as previously.  Within quoted string, it is possible to escape quote with `\"` or `\'`.

Note 1: An alternative approach would be to create new version of `flb_utils_split()` to make it absolutely sure the quotes are obeyed only for config files and not when splitting other input. This change currently does not do that and all test cases pass. 

Note 2: The current lack of support of quotes is not documented as far as I can find. I'm unsure if this change requires documentation.

Addresses #1225, #4286, #2415

----

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change
- [ ] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [N/A] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**

- [ ] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by tsaarni** [comment]
```console 
$ cat fluent-bit.conf
[INPUT]
    name tcp

[OUTPUT]
    name stdout

[FILTER]
    name modify
    match *
    condition key_value_matches message .*FOO.*
    add facility "audit log"

$ valgrind fluent-bit -v --config=fluent-bit.conf
==262966== Memcheck, a memory error detector
==262966== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==262966== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==262966== Command: fluent-bit -v --config=fluent-bit.conf
==262966== 
Fluent Bit v2.0.5
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/11/11 11:25:53] [ info] Configuration:
[2022/11/11 11:25:53] [ info]  flush time     | 1.000000 seconds
[2022/11/11 11:25:53] [ info]  grace          | 5 seconds
[2022/11/11 11:25:53] [ info]  daemon         | 0
[2022/11/11 11:25:53] [ info] ___________
[2022/11/11 11:25:53] [ info]  inputs:
[2022/11/11 11:25:53] [ info]      tcp
[2022/11/11 11:25:53] [ info] ___________
[2022/11/11 11:25:53] [ info]  filters:
[2022/11/11 11:25:53] [ info]      modify.0
[2022/11/11 11:25:53] [ info] ___________
[2022/11/11 11:25:53] [ info]  outputs:
[2022/11/11 11:25:53] [ info]      stdout.0
[2022/11/11 11:25:53] [ info] ___________
[2022/11/11 11:25:53] [ info]  collectors:
[2022/11/11 11:25:53] [ info] [fluent bit] version=2.0.5, commit=b4db5b6c67, pid=262966
[2022/11/11 11:25:53] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2022/11/11 11:25:53] [ info] [storage] ver=1.3.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2022/11/11 11:25:53] [ info] [cmetrics] version=0.5.7
[2022/11/11 11:25:53] [ info] [output:stdout:stdout.0] worker #0 started
[2022/11/11 11:25:53] [ info] [ctraces ] version=0.2.5
[2022/11/11 11:25:53] [ info] [input:tcp:tcp.0] initializing
[2022/11/11 11:25:53] [ info] [input:tcp:tcp.0] storage_strategy='memory' (memory only)
[2022/11/11 11:25:53] [debug] [tcp:tcp.0] created event channels: read=21 write=22
[2022/11/11 11:25:53] [debug] [downstream] listening on 0.0.0.0:5170
[2022/11/11 11:25:53] [debug] [filter:modify:modify.0] Creating regex for condition B : condition key_value_matches message .*FOO.* : .*FOO.*
[2022/11/11 11:25:53] [debug] [filter:modify:modify.0] Initialized modify filter with 1 conditions and 1 rules
[2022/11/11 11:25:53] [debug] [stdout:stdout.0] created event channels: read=24 write=25
[2022/11/11 11:25:53] [ info] [sp] stream processor started
[2022/11/11 11:25:56] [debug] [filter:modify:modify.0] Match for condition KEY_VALUE_MATCHES .*FOO.*
[2022/11/11 11:25:56] [debug] [input chunk] update output instances with new chunk size diff=56
[0] tcp.0: [1668158756.143888053, {"message"=>"Hello FOO world!", "facility"=>"audit log"}]
[2022/11/11 11:25:56] [debug] [task] created task=0x504a270 id=0 OK
[2022/11/11 11:25:56] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2022/11/11 11:25:56] [debug] [out flush] cb_destroy coro_id=0
[2022/11/11 11:25:56] [debug] [task] destroy task=0x504a270 (task_id=0)
^C[2022/11/11 11:26:01] [engine] caught signal (SIGINT)
[2022/11/11 11:26:01] [ warn] [engine] service will shutdown in max 5 seconds
[2022/11/11 11:26:01] [ info] [engine] service has stopped (0 pending tasks)
[2022/11/11 11:26:01] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/11/11 11:26:01] [ info] [output:stdout:stdout.0] thread worker #0 stopped
==262966== 
==262966== HEAP SUMMARY:
==262966==     in use at exit: 0 bytes in 0 blocks
==262966==   total heap usage: 1,639 allocs, 1,639 frees, 917,490 bytes allocated
==262966== 
==262966== All heap blocks were freed -- no leaks are possible
==262966== 
==262966== For lists of detected and suppressed errors, rerun with: -s
==262966== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)

# execute following on other terminal while fluent-bit is running to produce input for tcp input
$ echo '{ "message": "Hello FOO world!" }' > /dev/tcp/localhost/5170

**Comment by tsaarni** [comment]
Fuzzer seemed to [produce SIGSEGV](https://github.com/fluent/fluent-bit/actions/runs/3443974348/jobs/5746184138). I'm not sure how to interpret the output and how to reproduce.  I see there is base64 encoded fuzzed config file in the logs, so I tried following:

```console
$ echo W1BdU10KTG5hbWUgagpMdHlwZXMgJyVaCkxmb3JtYXQgbHRzdgo= | base64 -d > fuzzer.conf
$ cmake -DSANITIZE_ADDRESS=On ..
$ make
$ bin/fluent-bit --config=fuzzer.conf
```

but it did not reproduce the crash.

**Comment by tsaarni** [comment]
Just a friendly ping, any feedback for this PR would be greatly appreciated :)

**Comment by ekxxsho** [comment]
We have the same issue, we could not use the modify filter with any space between. This change is working for our scenario too. 
Is it possible for us to have this fix merged?

**Review by nokute78 — APPROVED** [review]
@tsaarni Thank you for contribution.

I approved since the unit tests that I added to test `flb_utils_split` passed.
https://github.com/fluent/fluent-bit/pull/6623

@edsiper Could you merge this ?

**Comment by tsaarni** [comment]
Hi @edsiper, when you have time, it would be greatly appreciated if you could have a look at this PR. Thank you!

**Comment by tsaarni** [comment]
Please hold, I will work some more with this: I will change the approach so that the behavior of split can be kept, and this change will only impact the current parser in modify plugin. 

Cc @leonardo-albertovich

**Comment by tsaarni** [comment]
I've now updated the PR by adding a separate function `flb_utils_split_quoted()` which parses quoted strings into tokens, while the existing `flb_utils_split()` keeps the old behavior.  

For now, only the modify plugin is calling the new `flb_utils_split_quoted()` to solve the issues addressed by this PR, but it may be useful for other purposes in the future as well.

**Comment by leonardo-albertovich** [comment]
I'm saving this PR now so I can properly review it and merge it on the 25th. Thank you for your patience.

**Review by leonardo-albertovich — CHANGES_REQUESTED** [review]

File: src/flb_utils.c [hunk 1]
```diff
@@ -257,7 +257,111 @@ void flb_utils_print_setup(struct flb_config *config)
...
+
+    token_out = flb_malloc(len + 1);
+    if (!token_out) {
+        return -1;
+    }
+
+    /* Copy the token */
+    for (i = 0; i < len;) {
+        /* Handle escapes */
+        if (*token_in == '\\' && (token_in[1] == quote || token_in[1] == '\\')) {
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply src/flb_utils.c:357]: True, thank you for spotting this! I've now fixed it and added test that would have caught it.

File: src/flb_utils.c [hunk 2]
```diff
@@ -257,7 +257,111 @@ void flb_utils_print_setup(struct flb_config *config)
...
+    /* Consume the quote character. */
+    token_in++;
+
+    token_out = flb_malloc(len + 1);
+    if (!token_out) {
+        return -1;
+    }
+
+    /* Copy the token */
+    for (i = 0; i < len;) {
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply src/flb_utils.c]: Fixed.

File: tests/internal/utils.c [hunk 3]
```diff
@@ -525,31 +532,41 @@ static int compare_split_entry(const char* input, int separator, int max_split,
...
+    compare_split_entry("aa,bb", '/', 2, FLB_FALSE, "aa,bb");
+
+    /* do not parse quotes when tokenizing */
+    compare_split_entry("aa \"bb cc\" dd", ' ', 256, FLB_FALSE, "aa", "\"bb", "cc\"", "dd");
+}
+
+void test_flb_utils_split_quoted()
+{
+    compare_split_entry("aa \"bb cc\" dd", ' ', 256, FLB_TRUE, "aa", "bb cc", "dd");
+    compare_split_entry("aa bb 'cc dd'", ' ', 256, FLB_TRUE, "aa", "bb",  "cc dd");
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply tests/internal/utils.c]: I've added tests.

**Review by leonardo-albertovich — DISMISSED** [review]

**Comment by leonardo-albertovich** [comment]
Great, I just approved the PR so CI can run, then I think you'll have to fix that DCO issue and we'll be able to merge it.

Thank you very much both for the contribution and the patience @tsaarni!

**Comment by tsaarni** [comment]
> Great, I just approved the PR so CI can run, then I think you'll have to fix that DCO issue and we'll be able to merge it.
> 
> Thank you very much both for the contribution and the patience @tsaarni!

Thank you @leonardo-albertovich for great review comments! Much appreciated! 

I force-pushed the last commit again to address the DCO sign-off issue. Could you accept the CI run again? Thank you!

**Comment by tsaarni** [comment]
@leonardo-albertovich Are the MacOS tests impacted by my change, or can it be unrelated flake?  

I cannot see the reason from [the logs](https://github.com/fluent/fluent-bit/actions/runs/4806465586/jobs/8556354237?pr=6387) other than

```
        Start   3: flb-rt-in_event_test
  1/120 Test   #3: flb-rt-in_event_test .......................Subprocess aborted***Exception:   4.86 sec
Test event_test...                              [2023/04/26 10:19:55] [ info] [fluent bit] version=2.1.2, commit=cb47b930b6, pid=20329
```

Some of the failures seem to have disappeared when you triggered a rerun.

**Comment by leonardo-albertovich** [comment]
Don't worry about those macos tests, the failures are not related to your PR and it's something we will fix in the near  future.

I'm tagging @edsiper here so he knows this is ready to be merged from my point of view.

**Comment by tsaarni** [comment]
Thank You @leonardo-albertovich for working with me on this PR, much appreciated! Hope to see some time again :smiley:

**Pull request merged**